### PR TITLE
ACNA-3231 | E2e tests failure

### PR DESCRIPTION
**Pull request by amulyakashyap09**
## Description

- a scenario where user does not have login but deployment creds
- if Audit service is unavailable then deploy/undeploy was failing

## Related Issue

[ACNA-3231](https://jira.corp.adobe.com/browse/ACNA-3231)

## How Has This Been Tested?

- Tested locally
- Unit Tests

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/lib/app-helper.js [hunk 1]
```diff
@@ -567,6 +570,30 @@ function getObjectValue (obj, key) {
...
 }
 
+/**
+ * Function to run `aio config get ims.contexts.cli.access_token.token` command and retrieve the token
+ * This function is used to check if the access token exists in the CLI context or not
+ * @returns {Promise<boolean>} Resolves to the access token if it exists, else false
+ */
+const checkifAccessTokenExists = async () => {
+  try {
+    const { stdout, stderr } = await execAsync('aio config get ims.contexts.cli.access_token.token')
```
Comment (shazron) [inline src/lib/app-helper.js:580]: use findCommand to double check command existence, see https://github.com/adobe/aio-cli-plugin-app/blob/9b36d06c0bc0614f0744aba87324cdf4c2b7059a/src/commands/app/pack.js#L165

Probably can use execa like in the above link.

File: src/lib/audit-logger.js [hunk 2]
```diff
@@ -33,7 +33,7 @@ const AUDIT_SERVICE_ENPOINTS = {
  * @param {string} env valid env stage|prod
  */
 async function sendAuditLogs (accessToken, logEvent, env = 'prod') {
-  const url = AUDIT_SERVICE_ENPOINTS[env]
+  const url = AUDIT_SERVICE_ENPOINTS[env] ?? AUDIT_SERVICE_ENPOINTS.prod
```
Comment (shazron) [inline src/lib/audit-logger.js:36]: spelling error `ENPOINTS`

File: src/commands/app/deploy.js [hunk 3]
```diff
@@ -52,7 +54,7 @@ class Deploy extends BuildCommand {
 
     try {
       const aioConfig = (await this.getFullConfig()).aio
-      const cliDetails = await getCliInfo()
+      const cliDetails = doesTokenExists ? await getCliInfo() : null
```
Comment (shazron) [inline src/commands/app/deploy.js:57]: for clarity I would name it `loginCredentials`

File: src/commands/app/deploy.js [hunk 4]
```diff
@@ -94,8 +96,12 @@ class Deploy extends BuildCommand {
 
       // 3. send deploy log event
       const logEvent = getAuditLogEvent(flags, aioConfig.project, 'AB_APP_DEPLOY')
-      if (logEvent) {
-        await sendAuditLogs(cliDetails.accessToken, logEvent, cliDetails.env)
+      if (logEvent && cliDetails) {
+        try {
+          await sendAuditLogs(cliDetails.accessToken, logEvent, cliDetails.env)
+        } catch (error) {
+          this.log(chalk.red(chalk.bold('Error: Audit Log Service Error: Failed to send audit log event for deployment.')))
```
Comment (shazron) [inline src/commands/app/deploy.js:103]: it is an error in sending, but to the cli is it an error? I would say it's more like a warning here. Just do a `this.warn` without all the `chalk` adornment

File: src/commands/app/deploy.js [hunk 5]
```diff
@@ -111,9 +117,13 @@ class Deploy extends BuildCommand {
...
           const opItems = getFilesCountWithExtension(v.web.distProd)
           const assetDeployedLogEvent = getAuditLogEvent(flags, aioConfig.project, 'AB_APP_ASSETS_DEPLOYED')
-          if (assetDeployedLogEvent) {
+          if (assetDeployedLogEvent && cliDetails) {
             assetDeployedLogEvent.data.opItems = opItems
-            await sendAuditLogs(cliDetails.accessToken, assetDeployedLogEvent, cliDetails.env)
+            try {
+              await sendAuditLogs(cliDetails.accessToken, logEvent, cliDetails.env)
+            } catch (error) {
+              this.log(chalk.red(chalk.bold('Error: Audit Log Service Error: Failed to send audit log event for deployment.')))
```
Comment (shazron) [inline src/commands/app/deploy.js:125]: it is an error in sending, but to the cli is it an error? I would say it's more like a warning here. Just do a `this.warn` without all the `chalk` adornment

File: src/commands/app/undeploy.js [hunk 6]
```diff
@@ -46,12 +47,16 @@ class Undeploy extends BaseCommand {
     const spinner = ora()
     try {
       const aioConfig = (await this.getFullConfig()).aio
-      const cliDetails = await getCliInfo()
+      const cliDetails = doesTokenExists ? await getCliInfo() : null
```
Comment (shazron) [inline src/commands/app/undeploy.js:50]: similarly, should be `loginCredentials` not `cliDetails`, for clarity

File: src/commands/app/undeploy.js [hunk 7]
```diff
@@ -46,12 +47,16 @@ class Undeploy extends BaseCommand {
...
       const logEvent = getAuditLogEvent(flags, aioConfig.project, 'AB_APP_UNDEPLOY')
 
       // 1.1. send audit log event for successful undeploy
-      if (logEvent) {
-        await sendAuditLogs(cliDetails.accessToken, logEvent, cliDetails.env)
+      if (logEvent && cliDetails) {
+        try {
+          await sendAuditLogs(cliDetails.accessToken, logEvent, cliDetails.env)
+        } catch (error) {
+          this.log(chalk.red(chalk.bold('Error: Audit Log Service Error: Failed to send audit log event for deployment.')))
```
Comment (shazron) [inline src/commands/app/undeploy.js:58]: it is an error in sending, but to the cli is it an error? I would say it's more like a warning here. Just do a `this.warn` without all the `chalk` adornment

Typo: should be `undeployment` (is that even a word?)

File: src/commands/app/undeploy.js [hunk 8]
```diff
@@ -61,8 +66,12 @@ class Undeploy extends BaseCommand {
         const v = values[i]
         await this.undeployOneExt(k, v, flags, spinner)
         const assetUndeployLogEvent = getAuditLogEvent(flags, aioConfig.project, 'AB_APP_ASSETS_UNDEPLOYED')
-        if (assetUndeployLogEvent) {
-          await sendAuditLogs(cliDetails.accessToken, assetUndeployLogEvent, cliDetails.env)
+        if (assetUndeployLogEvent && cliDetails) {
+          try {
+            await sendAuditLogs(cliDetails.accessToken, assetUndeployLogEvent, cliDetails.env)
+          } catch (error) {
+            this.log(chalk.red(chalk.bold('Error: Audit Log Service Error: Failed to send audit log event for deployment.')))
```
Comment (shazron) [inline src/commands/app/undeploy.js:73]: it is an error in sending, but to the cli is it an error? I would say it's more like a warning here. Just do a `this.warn` without all the `chalk` adornment.

Typo: should be `undeployment` (is that even a word?)

File: test/commands/app/undeploy.test.js [hunk 9]
```diff
@@ -531,4 +564,40 @@ describe('run', () => {
...
+      message: 'Internal Server Error',
+      status: 500
+    })
+
+    await command.run()
+    expect(command.error).toHaveBeenCalledTimes(0)
+    expect(mockRuntimeLib.undeployActions).toHaveBeenCalledTimes(1)
+    expect(mockWebLib.undeployWeb).toHaveBeenCalledTimes(1)
+    expect(auditLogger.sendAuditLogs).toHaveBeenCalledTimes(2)
+    expect(command.log).toHaveBeenCalledWith(expect.stringMatching('Error: Audit Log Service Error: Failed to send audit log event for deployment.'))
```
Comment (shazron) [inline test/commands/app/undeploy.test.js:601]: test failure if typo is fixed (see previous comments)

File: test/commands/lib/audit-logger.test.js [hunk 10]
```diff
@@ -90,6 +90,21 @@ test('sendAuditLogs with default params', async () => {
...
+    method: 'POST',
+    headers: {
+      Authorization: 'Bearer ' + mockToken,
+      'Content-type': 'application/json'
+    },
+    body: JSON.stringify({ event: mockLogEvent })
+  }
+  await auditLogger.sendAuditLogs(mockToken, mockLogEvent, 'dev')
+  expect(fetch).toHaveBeenCalledTimes(1)
+  expect(fetch).toHaveBeenCalledWith(auditLogger.AUDIT_SERVICE_ENPOINTS.prod, options)
```
Comment (shazron) [inline test/commands/lib/audit-logger.test.js:105]: test failure if typo is fixed (see previous comments)

**Review by shazron — CHANGES_REQUESTED** [review]

File: src/lib/app-helper.js (hunk 1 above)
Comment (MichaelGoberling) [reply src/lib/app-helper.js:580]: Is using execa the new way to run other commands or should we use `.load()` and `.run()` like here

https://github.com/adobe/aio-cli-plugin-app/blob/251f7bb62424d52d9eef307a161d28b3ddcc9d9f/src/commands/app/run.js#L154-L158

Comment (shazron) [reply src/lib/app-helper.js]: @MichaelGoberling actually those interfaces are not documented and we have to read the code to see how to do it. Perhaps they may go away? Unlikely, but I didn't even know about this interface. Even `findCommand` is not documented anymore, and `runCommand` is there instead: https://oclif.io/docs/config (not that they document how to use it!).

Not sure how that interface works if we need stdout/stderr.

Edit: I just saw they just have this guide: https://oclif.io/docs/running_programmatically/ which uses runCommand

**Review by shazron — APPROVED** [review]

File: src/lib/app-helper.js [hunk 11]
```diff
@@ -567,6 +567,29 @@ function getObjectValue (obj, key) {
...
 }
 
+/**
+ * Function to run `aio config get ims.contexts.cli.access_token.token` command and retrieve the token
+ * This function is used to check if the access token exists in the CLI context or not
+ * @returns {Promise<boolean>} Resolves to the access token if it exists, else false
+ */
+const checkifAccessTokenExists = async () => {
+  try {
+    const { stdout, stderr } = await execa('aio', ['config', 'get', 'ims.contexts.cli.access_token.token'])
```
Comment (purplecabbage) [inline src/lib/app-helper.js:577]: do not go through all this extra ceremony of calling the cli for another command just to read config.  The config is likely already loaded and in mem, even this file has other references to reading values from config without extra trips + exec.
Look at ln:21 of this file.

File: src/lib/audit-logger.js [hunk 12]
```diff
@@ -16,12 +16,12 @@ const chalk = require('chalk')
...
-  AB_APP_TEST: 'ab_app_test', // todo : remove after testing
+  AB_APP_TEST: 'ab_app_test',
   AB_APP_ASSETS_DEPLOYED: 'ab_app_assets_deployed',
   AB_APP_ASSETS_UNDEPLOYED: 'ab_app_assets_undeployed'
 }
 
-const AUDIT_SERVICE_ENPOINTS = {
+const AUDIT_SERVICE_ENDPOINTS = {
   stage: 'https://adp-auditlog-service-stage.adobeioruntime.net/api/v1/web/audit-log-api/event-post',
   prod: 'https://adp-auditlog-service-prod.adobeioruntime.net/api/v1/web/audit-log-api/event-post'
```
Comment (purplecabbage) [inline src/lib/audit-logger.js:26]: If this endpoint is not live it should never appear in a code commit.

File: test/commands/app/deploy.test.js [hunk 13]
```diff
@@ -279,7 +280,7 @@ describe('run', () => {
     command.appConfig = cloneDeep(mockConfigData)
     command.appConfig.actions = { dist: 'actions' }
     command.appConfig.web.distProd = 'dist'
-    command.config = { runCommand: jest.fn(), runHook: jest.fn() }
+    command.config = { runCommand: jest.fn(), runHook: jest.fn(), findCommand: jest.fn().mockReturnValue({}) }
```
Comment (purplecabbage) [inline test/commands/app/deploy.test.js:283]: none of this is needed

**Review by purplecabbage — CHANGES_REQUESTED** [review]
Skip all this findCommand, execa call command work, it is not required, just use aioConfig directly as is already done throughout the code.

**Review by arjuncooliitr — APPROVED** [review]

**Pull request closed without merging**